### PR TITLE
Feature/ted 87 metadata normaliser - extractor

### PR DESCRIPTION
**Pull request by Dragos0000**

### CODE REVIEW

File: tests/unit/notice_normaliser/test_metadata_extractor.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    assert extracted_metadata_dict.keys() == ExtractedMetadata.__fields__.keys()
+    assert notice_2015.ted_id in extracted_metadata_dict["notice_publication_number"]
+
+def test_metadata_extractor_2018(notice_2018):
+    metadata_extractor = MetadataExtractor(notice=notice_2018).extract_metadata()
+
+    extracted_metadata_dict = metadata_extractor.dict()
+    assert isinstance(metadata_extractor, ExtractedMetadata)
+    assert extracted_metadata_dict.keys() == ExtractedMetadata.__fields__.keys()
+    assert notice_2018.ted_id in extracted_metadata_dict["notice_publication_number"]
```
Comment (costezki) [inline tests/unit/notice_normaliser/test_metadata_extractor.py:42]: why only this key? 

You see, This is the reason why I recommended to list the keys as variables in the "feature file"

File: tests/unit/notice_normaliser/test_metadata_normaliser.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from ted_sws.domain.model.notice import NoticeStatus
+from ted_sws.metadata_normaliser.services.metadata_normalizer import MetadataNormaliser
+
+
+def test_metadata_extractor(raw_notice):
```
Comment (costezki) [inline tests/unit/notice_normaliser/test_metadata_normaliser.py:5]: this might be a bit more sophisticated

**Review by costezki — CHANGES_REQUESTED** [review]
I like your efforts in figuring out XPaths! Also good start in organising the domain and service layers. We need some refactoring and simplification.

File: ted_sws/metadata_normaliser/services/__init__.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+from xml.etree.ElementTree import Element
+
+from ted_sws.domain.model.metadata import ExtractedMetadata
+
+
+def xpath_extract_data(elements: List[Element]) -> List[str]:
+    return [element.text for element in elements]
+
+
+def xpath_extract_attributes(elements: List[Element], attrib_key: str) -> List[str]:
```
[the remark this answers is not in the document]
Comment (Dragos0000) [reply ted_sws/metadata_normaliser/services/__init__.py:11]: renamed

File: ted_sws/metadata_normaliser/services/__init__.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    return [element.attrib[attrib_key] for element in elements]
+
+
+class ExtractMetadata:
```
[the remark this answers is not in the document]
Comment (Dragos0000) [reply ted_sws/metadata_normaliser/services/__init__.py:15]: moved and renamed

File: tests/conftest.py [hunk 5]
```diff
@@ -12,4 +19,58 @@ def notice_repository():
 
 @pytest.fixture
 def ted_document_search():
-    return TedAPIAdapter(request_api=FakeRequestAPI())
\ No newline at end of file
+    return TedAPIAdapter(request_api=FakeRequestAPI())
```
[the remark this answers is not in the document]
Comment (Dragos0000) [reply tests/conftest.py:22]: for  testing with fakes

File: tests/conftest.py [hunk 6]
```diff
@@ -1,9 +1,16 @@
+import base64
```
[the remark this answers is not in the document]
Comment (Dragos0000) [reply tests/conftest.py:1]: we need to decode the content ... as a notice for us is a api response

File: ted_sws/metadata_normaliser/services/xml_manifestation_metadata_extractor.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+
+
+def extract_code_and_value_from_element(element: ET.Element) -> EncodedValue:
+    """
+    Extract code attribute and text values from an element in the XML structure
+    :param element:
+    :return:
+    """
+    if element is not None:
+        return EncodedValue(code=extract_attribute_from_element(element=element, attrib_key="CODE"),
```
Comment (costezki) [inline ted_sws/metadata_normaliser/services/xml_manifestation_metadata_extractor.py:39]: Warning: we might not always have "CODE" attribute; it is worth having it as a default variable.

**Review by costezki — APPROVED** [review]
It is good to go after all the previous comments are addressed.

**Pull request merged**